README.md
# Stockpile

Stockpile is a plugin for MODX Revolution that caches on save. Also provides a snippet, getStockpile
to get resource from cache, no DB/xPDO queries. Saves all fields and TV values to cache. It can also save resources as static.
Similar to the StatCache extra.

## Install via Composer

Follow the [Local Orchestrator](https://github.com/LippertComponents/LocalOrchestrator) helper to place in the correct 
location and then edit the composer.json file. Place "lci/stockpile": "dev-master" in the require spot. And then 
put "lci/stockpile" in the auto-install array. And then run composer install.

## What is installed?

### Snippet

**getStockpile**

Use the getStockpile snippet to process a chunk form cached resource data or it can send the cached resource 
fields/tvs data to placeholders

| Property | Description | Default |
|--- |--- |--- |
| resourceID| int required,  a valid resource ID | |
| item | string chunk name to be be processed with resource data. If empty snippet will return empty ex: &item=`myChunk` | |
| sendToPlaceholders | boolean - send the data to placeholders | 1 |
| prefix | string - for placeholders, example: &prefix=`sp` [[+sp.pagetitle]] | |
| published | bool - require a resource to be published to complete request, if 0 then will show all | 1 |
| debug | bool - show property values | 0 |

### Plugin

stockpile plugin is installed and is will save data on the following events

| Event | Description |
| --- | --- |
| OnBeforeSaveWebPageCache | It will do $staticGenerator->makeResourceStaticFileOnWebCache($modx->resource); is related option is true |
| OnResourceAutoPublish | It will cache every published resource and remove cache for every unpublished resource |
| OnDocFormSave, OnDocPublished, OnResourceUndelete | It will cache resource |
| OnDocUnPublished | Remove cache |
| OnResourceDelete | Delete resource cache if remaining |
| OnSiteRefresh | Rebuild static cache if LCI_STOCKPILE_REGENERATE_ON_CLEAR_CACHE=1 and static cache is enabled |
| FredOnFredResourceSave | If you use Fred, you will need to manually attach this event to the stockpile plugin after you have installed Fred |

### xPDO, database table

stockpile_cache_que table with class name StockpileQueLog

### Events to write custom plugins

Easily customize what is cached on saved via the OnStockpileSave event. Write your own plugin.

`OnStockpileSave` event: 
    - @var \LCI\MODX\Stockpile $stockpile
    - @var \modResource $resource
    - @var array $data - resource fields and TVs
    - can use $stockpile->setResourceData($data); after you make modifications to save the data to the cache file

`OnStockpileAfterSaveMakeQueLog` and `OnStockpileAfterDeleteMakeQueLog`

#### Example to create a cache buster

```php
use LCI\MODX\Stockpile\Stockpile;
use LCI\MODX\Stockpile\StockpileQue;

$eventName = $modx->event->name;

/**
 * This plugin should only be called in the OnStockpileAfterSaveMakeQueLog and OnStockpileAfterDeleteMakeQueLog event
 * LCI1.com custom plugin to bust the stockpile static cache for related resources
 *
 * @param Stockpile $stockpile
 * @param StockpileQue $stockpileQue
 * @param modResource $resource
 * @param array $data - the resource data as a stockpile array
 */

switch($eventName) {
    case 'OnStockpileAfterDeleteMakeQueLog':
        // no break
    case 'OnStockpileAfterSaveMakeQueLog':
        if ($resource->get('id') == 10) {
            $stockpileQue->rebuildAll();
        }
        
        // Blog:
        $stockpileQue->rebuildParentWhenResourceHasParentID($resource, 10);

        // Products that can show up on a lot of product category pages
        $parent = $resource->get('parent');
        if (in_array($parent, [50, 75])) {
            $stockpileQue->rebuildResourceIDs([20, 22, 30, 45, 63]);
        }

        // Siblings
        if ($resource->get('id') == 15) {
            $stockpileQue->rebuildResourceID(16);
        }
        if ($resource->get('id') == 16) {
            $stockpileQue->rebuildResourceID(15);
        }
        break;
}
```

## CLI Command line

 - ```cd /www/core/vendor/bin```
 - ```php orchestrator``` this will show a complete list of options
 
 _Note_ 
You can run the orchestrator commands from anywhere in the path. For example on MODXCloud this works:  
```php /www/core/vendor/bin/orchestrator ``` 
AND
```
cd /www/core/
php vendor/bin/orchestrator 
```
 
### Stockpile commands
 
 - `php orchestrator stockpile:build` ~ this will (re)cache all resources
 - `php orchestrator stockpile:build --ids 2,3` ~ this will (re)cache a comma separated list of resource IDs
    - Short hand: `php orchestrator stockpile:build -i 2,3`
 - `php orchestrator stockpile:remove` ~ Clear/Remove all stockpile cache
 - `php orchestrator stockpile:remove --ids 2,3` ~ clear/remove cache for that resource ID
    - Short hand: `php orchestrator stockpile:remove -i 2,3`
 - `php orchestrator stockpile:que`
    Process Stockpile cache que, if using a custom plugin then it is suggested to set a cron job to run this process 
    every 5 to 15 minutes depending on how fast you want related content pages to be recached.

## Static Generator 

Based on StatCache: https://github.com/opengeek/statcache, configure with the .env settings below.

### Nginx rules

Example on MODXCloud

*Remove*
```
location / {
    try_files $uri $uri/ @modx-rewrite;
}
```

*Replace with*
```
# Start Stockpile Static Generator
set $cache_prefix 'core/cache/static';
if ($http_user_agent = 'MODX RegenCache') {
    set $cache_prefix 'no cache';
}

location / {
    try_files /$cache_prefix$uri~index.html /$cache_prefix$uri~.html /$cache_prefix$uri $uri $uri/ @modx-rewrite;
}

# END Stockpile Static Generator
```

## Configure with .env

If you have not created an .env in the MODX/core directory do so and optionally add the following

| Name | Default | Description | 
| --- | --- | --- |
| LCI_STOCKPILE_ENABLE_STATIC | 0 | Set to 1 or true to enable caching resources to static | 
| LCI_STOCKPILE_CACHE_PATH | core/cache/static/ | The path to cache static resources appended to MODX_BASE_PATH | 
| LCI_STOCKPILE_EXCLUDE_TV_NAME |  | Optional, if set and TV exists and resource has a value of 1 or true it will not be cached. | 
| LCI_STOCKPILE_CONTENT_TYPES |  | If specified and non-empty, only cache Resources with the specified ContentType id's. Accepts a comma-delimited list of ContentType id's. | 
| LCI_STOCKPILE_CONTEXTS |  | If specified and non-empty, only cache Resources in the specified Contexts. Accepts a comma-delimited list of Context keys. | 
| LCI_STOCKPILE_MIME_TYPES |  | If specified and non-empty, only cache Resources with the specified mime-types. Accepts a comma-delimited list of mime-types. | 
| LCI_STOCKPILE_EXCLUDE_BINARY_CONTENT_TYPES | 1 | Skip Resources that have a binary Content Type. | 
| LCI_STOCKPILE_EXCLUDE_REMAINING_TAGS | 1 | Exclude Resources that have tags remaining in the content that is being cached for the Resource. | 
| LCI_STOCKPILE_REGENERATE_ON_CLEAR_CACHE | 0 | If 1 will attempt to regenerate static files when clearing site cache. | 
| LCI_STOCKPILE_REGENERATE_ON_SAVE | 1 | Regenerate an existing static file when a Resource is saved in the manager. | 
| LCI_STOCKPILE_REGENERATE_USERAGENT | MODX RegenCache | The User-Agent HTTP header to send when regenerating static files. Your web server should be configured to not serve the static files when the User-Agent equals the value specified here. | 
| LCI_STOCKPILE_USE_URL_SCHEME |  | If enabled, includes the url_scheme (without the ://) as part of the `statcache_path`. Useful for sites using multiple Contexts for sub-domains or multiple domains. NOTE: using this requires changes to your web server rewrites rules. | 
| LCI_STOCKPILE_USE_HTTP_HOST |  | If enabled, includes the http_host as part of the `statcache_path`. Useful for sites using multiple Contexts for sub-domains or multiple domains. NOTE: using this requires changes to your web server rewrites rules. | 




CHANGELOG.md
# Stockpile - Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
Given a version number MAJOR.MINOR.PATCH, increment the:

    MAJOR version when you make incompatible site breaking changes,
    MINOR version when you add functionality in a backwards-compatible manner, and
    PATCH version when you make backwards-compatible bug fixes.

## [v1.4.1](https://github.com/LippertComponents/Stockpile/compare/v1.4.0...v1.4.1) - 2019-08-21
### Changed
- Listen to the OnEmptyTrash event

## [v1.4.0](https://github.com/LippertComponents/Stockpile/compare/v1.3.3...v1.4.0) - 2019-08-14
### Added
- Add switch for event FredOnFredResourceSave on Stockpile plugin, will need to manually attach event to plugin after Fred has been installed.

## [v1.3.3](https://github.com/LippertComponents/Stockpile/compare/v1.3.2...v1.3.3) - 2019-08-13
### Changed
- Added new option --exclude to exclude resource children for building with passing the parent ID

## [v1.3.2](https://github.com/LippertComponents/Stockpile/compare/v1.3.1...v1.3.2) - 2019-06-21
### Changed
- Fix caching modWeblink to set the content and full_url as the processed/parsed content to make the destination URL 

## [v1.3.1](https://github.com/LippertComponents/Stockpile/compare/v1.3.0...v1.3.1) - 2019-06-11
### Changed
- Fix stockpile:build and stockpile:remove to do all resources with no options as expected

## [v1.3.0](https://github.com/LippertComponents/Stockpile/compare/v1.2.1...v1.3.0) - 2019-06-07
### Added 
- stockpile:que command to allow running a cron job to rebuild cache of only those in the que
- StockpileQue class
- New Events: OnStockpileAfterSaveMakeQueLog and OnStockpileAfterDeleteMakeQueLog, you can create a custom plugin to write 
your own cache buster strategy and maintain static generated pages
- Stockpile->removeResourceCache will now delete the MODX Resource cache as well as the stockpile cache

### Changed
- Simplified the build and remove command options

## [v1.2.1](https://github.com/LippertComponents/Stockpile/compare/v1.2.0...v1.2.1) - 2019-05-30
### Added 
- Add missing run stats to the Build and Remove Cache commands

## [v1.2.0](https://github.com/LippertComponents/Stockpile/compare/v1.1.1...v1.2.0) - 2019-05-30
### Added 
-  Add makeRemainingTagsUncacheableStaticFileOnWebCache method that will write the content to a -remaining-tags directory so they can be reviewed

### Changed
- Fix RemoveCache command to send resource ID not object

## [v1.1.1](https://github.com/LippertComponents/Stockpile/compare/v1.0.1...v1.1.0) - 2019-05-30
### Changed
- Fix AddStaticGenerator migration to overwrite existing plugin 

## [v1.1.0](https://github.com/LippertComponents/Stockpile/compare/v1.0.1...v1.1.0) - 2019-05-29
### Added
- Add in StaticGenerator and migration to attach events to the Stockpile plugin, configure with .env file, see readme

## [v1.0.1](https://github.com/LippertComponents/Stockpile/compare/v1.0.0...v1.0.1) - 2019-04-25
### Change
-  Fix caching Tagger Tags when the Group name is different that the group alias

## [v1.0.0](https://github.com/LippertComponents/Stockpile/releases/tag/v1.0.0) - 2019-03-23
### First stable release


src/StaticGenerator.php
<?php
/**
 * Created by PhpStorm.
 * User: joshgulledge
 * Date: 2019-03-15
 * Time: 06:17
 */

namespace LCI\MODX\Stockpile;

use modResource;
use modX;
use Symfony\Component\Console\Style\SymfonyStyle;

class StaticGenerator
{
    /** @var  \modX */
    protected $modx;

    /** @var \modCacheManager */
    protected $cacheManager;

    /** @var array  */
    protected $config = [];

    /**
     * StaticGenerator constructor.
     * @param $modx
     */
    public function __construct($modx)
    {
        $this->modx = $modx;
        $this->cacheManager = $modx->getCacheManager();

        $config = getenv();

        $this->config = [
            'LCI_STOCKPILE_ENABLE_STATIC' => (bool)$this->modx->getOption('LCI_STOCKPILE_ENABLE_STATIC', $config, true),
            'LCI_STOCKPILE_CACHE_PATH' => $this->modx->getOption('LCI_STOCKPILE_CACHE_PATH', $config, MODX_BASE_PATH .'core/cache/static/'),
            'LCI_STOCKPILE_CACHE_TV_NAME' => $this->modx->getOption('LCI_STOCKPILE_CACHE_TV_NAME', $config, ''),
            'LCI_STOCKPILE_CONTENT_TYPES' => $this->getAsArray($this->modx->getOption('LCI_STOCKPILE_CONTENT_TYPES', $config, '')),
            'LCI_STOCKPILE_CONTEXTS' => $this->getAsArray($this->modx->getOption('LCI_STOCKPILE_CONTEXTS', $config, '')),
            'LCI_STOCKPILE_MIME_TYPES' => $this->getAsArray($this->modx->getOption('LCI_STOCKPILE_MIME_TYPES', $config, '')),
            'LCI_STOCKPILE_EXCLUDE_BINARY_CONTENT_TYPES' => (bool)$this->modx->getOption('LCI_STOCKPILE_EXCLUDE_BINARY_CONTENT_TYPES', $config, true),
            'LCI_STOCKPILE_EXCLUDE_REMAINING_TAGS' => (bool)$this->modx->getOption('LCI_STOCKPILE_EXCLUDE_REMAINING_TAGS', $config, true),
            'LCI_STOCKPILE_REGENERATE_ON_CLEAR_CACHE' => (bool)$this->modx->getOption('LCI_STOCKPILE_REGENERATE_ON_CLEAR_CACHE', $config, false),
            'LCI_STOCKPILE_REGENERATE_ON_SAVE' =>(bool) $this->modx->getOption('LCI_STOCKPILE_REGENERATE_ON_SAVE', $config, true),
            'LCI_STOCKPILE_REGENERATE_USERAGENT' => $this->modx->getOption('LCI_STOCKPILE_REGENERATE_USERAGENT', $config, 'MODX RegenCache'),
            'LCI_STOCKPILE_USE_URL_SCHEME' => (bool)$this->modx->getOption('LCI_STOCKPILE_USE_URL_SCHEME', $config, false),
            'LCI_STOCKPILE_USE_HTTP_HOST' => (bool)$this->modx->getOption('LCI_STOCKPILE_USE_HTTP_HOST', $config, false)
        ];
    }

    /**
     * @return mixed
     */
    public function deleteAllStaticResourcesFiles()
    {
        $this->cacheManager->deleteTree(
            rtrim($this->config['LCI_STOCKPILE_CACHE_PATH'], '/').'-remaining-tags/',
            [
                'deleteTop' => false,
                'skipDirs' => false,
                'extensions' => [],
            ]
        );

        return $this->cacheManager->deleteTree(
            $this->config['LCI_STOCKPILE_CACHE_PATH'],
            [
                'deleteTop' => false,
                'skipDirs' => false,
                'extensions' => [],
                // #20:
                //'delete_exclude_items' => explode(',', $modx->getOption('statcache_delete_exclude', $scriptProperties, ''))
            ]
        );
    }

    /**
     * @param modResource $resource
     * @return bool
     */
    public function deleteStaticResourceFile(modResource $resource)
    {
        $deleted = false;
        $static_resource_file = $this->getStaticResourcePath($resource);

        if (is_readable($static_resource_file)) {
            if ($deleted = unlink($static_resource_file) === false) {
                $this->modx->log(modX::LOG_LEVEL_ERROR, "Error removing static file {$static_resource_file}", '', __METHOD__);
            }
        }

        $static_remaining_tags_resource_file = $this->getStaticResourcePath($resource, true);

        if (is_readable($static_remaining_tags_resource_file)) {
            unlink($static_remaining_tags_resource_file);
        }

        return $deleted;
    }

    /**
     * @param modResource $resource
     */
    public function rebuildStaticResource(modResource $resource)
    {
        $this->callResourceViaHTTPToCache($resource);
    }

    /**
     * @param modResource $resource
     */
    public function rebuildStaticResourceOnSave(modResource $resource)
    {
        if ($this->config['LCI_STOCKPILE_REGENERATE_ON_SAVE']) {
            $this->callResourceViaHTTPToCache($resource);
        }
    }

    /**
     *
     */
    public function rebuildAllResourcesOnClearCache()
    {
        if ($this->config['LCI_STOCKPILE_REGENERATE_ON_CLEAR_CACHE']) {

            $collection = $this->getEligibleResources();

            foreach ($collection as $resource) {
                $this->callResourceViaHTTPToCache($resource);
            }
        }
    }

    /**
     * @param SymfonyStyle $symfonyStyle
     * @return int
     */
    public function rebuildAllResources(SymfonyStyle $symfonyStyle)
    {
        $collection = $this->getEligibleResources();

        $count = 0;
        $total = count($collection);
        $symfonyStyle->progressStart($total);

        foreach ($collection as $resource) {
            if ($this->callResourceViaHTTPToCache($resource)) {
                $count++;
            }

            $symfonyStyle->progressAdvance();
        }

        $symfonyStyle->progressFinish();

        return $count;
    }

    /**
     * @param modResource $resource
     * @return bool
     */
    public function makeResourceStaticFileOnWebCache(modResource $resource)
    {
        if (!$this->config['LCI_STOCKPILE_ENABLE_STATIC'] || !$this->canSaveAsStatic($resource) || $resource->_output == '') {
            return false;
        }

        $static_resource_file = $this->getStaticResourcePath($resource);

        /* attempt to write the complete Resource output to the static file */
        if (!$written = $this->modx->cacheManager->writeFile($static_resource_file, $resource->_output)) {
            $this->modx->log(modX::LOG_LEVEL_ERROR, "Error caching output from Resource {$resource->get('id')} to static file {$static_resource_file}", '', __METHOD__);
        }

        return $written;
    }

    /**
     * @param modResource $resource
     * @param bool $add_remaining_tags
     * @return string
     */
    public function getStaticResourcePath(modResource $resource, $add_remaining_tags=false)
    {
        if ($resource->Context->config === null) {
            $resource->Context->prepare();
        }

        $resourceContext = $resource->Context;


        $path = rtrim($this->config['LCI_STOCKPILE_CACHE_PATH'], '/').($add_remaining_tags ? '-remaining-tags' : '').'/';

        /* generate an absolute URI representation of the Resource to append to the path */
        if ($resource->get('id') === (integer)$resourceContext->getOption('site_start', 1)) {
            $uri  = $resourceContext->getOption('base_url', MODX_BASE_URL);
            /* use ~index.html to represent the site_start Resource */
            $uri .= '~index.html';

        } else {
            $uri = $this->modx->makeUrl($resource->get('id'), $resource->get('context_key'), '', 'abs');
            if (strpos($uri, $resourceContext->getOption('url_scheme') . $resourceContext->getOption('http_host')) === 0) {
                /* remove url_scheme and http_host from any full URLs generated by MODX automatically */
                $uri = substr($uri, strlen($resourceContext->getOption('url_scheme') . $resourceContext->getOption('http_host')));

            } elseif (!empty($resourceContext->getOption('site_url')) && strpos($uri, $resourceContext->getOption('site_url')) === 0) {
                /* remove url_scheme and http_host from any full URLs generated by MODX automatically */
                $uri = substr($uri, strlen($resourceContext->getOption('site_url')));
            }

            if (substr($uri, strlen($uri) - 1) === '/' && $resource->ContentType->get('mime_type') == 'text/html') {
                /* if Resource is HTML and ends with a /, use ~index.html for the filename */
                $uri .= '~index.html';

            } elseif ($resource->ContentType->get('mime_type') == 'text/html' && $resource->ContentType->get('file_extensions') == '' ) {
                $uri .= '~.html';
            }

        }

        if ($this->config['LCI_STOCKPILE_USE_URL_SCHEME']) {
            $path .= str_replace('://', '', $resourceContext->getOption('url_scheme')).'/';
        }
        if ($this->config['LCI_STOCKPILE_USE_HTTP_HOST']) {
            $path .= rtrim($resourceContext->getOption('http_host'), '/').'/';
        }

        return $path . ltrim($uri, '/');
    }

    /**
     * @param modResource $resource
     * @return bool
     */
    protected function canSaveAsStatic(modResource $resource)
    {
        if ($resource->get('deleted') || !$resource->get('cacheable') || !$resource->get('published') || $resource->get('class_key') == 'modWebLink') {
            return false;
        }

        if (!empty($this->config['LCI_STOCKPILE_EXCLUDE_TV_NAME']) && (bool)$resource->getTVValue($this->config['LCI_STOCKPILE_EXCLUDE_TV_NAME'])) {
            return false;
        }

        /* optionally skip binary content types */
        if ($this->config['LCI_STOCKPILE_EXCLUDE_BINARY_CONTENT_TYPES'] && $resource->ContentType->get('binary')) {
            return false;
        }

        /* do not cache if the cacheable content still contains unprocessed tags */
        $matches = [];

        if ($this->config['LCI_STOCKPILE_EXCLUDE_REMAINING_TAGS'] &&
            $this->modx->parser->collectElementTags($resource->_content, $matches)) {
            $this->makeRemainingTagsUncacheableStaticFileOnWebCache($resource);
            return false;
        }

        /* if specified, limit caching by mime-type */
        if (count($this->config['LCI_STOCKPILE_MIME_TYPES']) &&
            !in_array(strtolower($resource->ContentType->get('mime_type')), $this->config['LCI_STOCKPILE_MIME_TYPES'])) {
            return false;
        }

        /* if specified, limit caching by ContentTypes */
        if (count($this->config['LCI_STOCKPILE_CONTENT_TYPES']) &&
            !in_array($resource->ContentType->get('id'), $this->config['LCI_STOCKPILE_CONTENT_TYPES'])) {
            return false;
        }

        if (count($this->config['LCI_STOCKPILE_CONTEXTS']) && !in_array($resource->get('context_key'), $this->config['LCI_STOCKPILE_CONTEXTS'])) {
            return false;
        }

        return true;
    }

    /**
     * @param modResource $resource
     * @return bool
     */
    protected function callResourceViaHTTPToCache(modResource $resource)
    {
        $this->deleteStaticResourceFile($resource);

        if ($this->config['LCI_STOCKPILE_ENABLE_STATIC'] && $this->canSaveAsStatic($resource)) {
            /**
             * @TODO Guzzle:
             */
            $curl = curl_init();
            curl_setopt($curl, CURLOPT_USERAGENT, $this->config['LCI_STOCKPILE_REGENERATE_USERAGENT']);
            curl_setopt($curl, CURLOPT_FAILONERROR, false);
            curl_setopt($curl, CURLOPT_FOLLOWLOCATION, false);
            curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
            curl_setopt($curl, CURLOPT_NOBODY, true);
            $url = $this->modx->makeUrl($resource->get('id'), $resource->get('context_key'), '', 'full');
            if (!empty($url)) {
                $this->modx->log(modX::LOG_LEVEL_INFO, "Requesting Resource at {$url}");
                curl_setopt($curl, CURLOPT_URL, $url);
                curl_exec($curl);
                $this->modx->log(modX::LOG_LEVEL_INFO, "Updated cache for resource at {$url}");
            }
            curl_close($curl);

            return true;
        }

        return false;
    }

    /**
     * @param string $input
     * @return array|bool
     */
    protected function getAsArray($input)
    {
        $input = explode(',', $input);
        if ($input === false) {
            return [];
        }
        if (is_array($input)) {
            array_walk($input, 'trim');
        }

        return [];
    }
    /**
     * @return boolean|array
     */
    protected function getEligibleResources()
    {
        //$this->deleteAllStaticResourcesFiles();

        $query = $this->modx->newQuery('modResource');
        $query->where([
            'cacheable' => true,
            'class_key:!=' => 'modWebLink',
            'deleted' => 0,
            'published' => 1,
        ]);

        return $this->modx->getCollection('modResource', $query);
    }

    /**
     * @param modResource $resource
     * @return mixed
     */
    public function makeRemainingTagsUncacheableStaticFileOnWebCache(modResource $resource)
    {
        $static_resource_file = $this->getStaticResourcePath($resource, true);

        /* attempt to write the complete Resource output to the static file */
        if (!$written = $this->modx->cacheManager->writeFile($static_resource_file, $resource->_content)) {
            $this->modx->log(modX::LOG_LEVEL_ERROR, "Error caching _content from Resource {$resource->get('id')} to static file {$static_resource_file}", '', __METHOD__);
        }

        return $written;
    }

}